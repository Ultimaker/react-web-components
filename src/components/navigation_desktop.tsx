import * as React from 'react';
import { NavLink, withRouter, RouteComponentProps } from 'react-router-dom';

// components
import SubNavigationMenu from './sub_navigation_menu';
import SubNavigationMenuItem from './sub_navigation_menu_item';
import { NavRoute } from './navigation';

export interface NavigationDesktopProps extends RouteComponentProps<{}> {
    /** Details for the visible nav items */
    visibleNavLinks: NavRoute[];
    /** A label to be shown in the navigation bar */
    navLabel?: string;
}


export const NavigationDesktop: React.StatelessComponent<NavigationDesktopProps> = ({
    navLabel, visibleNavLinks, location,
}) => (
    <div className="navigation__desktop show-sm layout layout--gutter-none layout--align-middle">
        {visibleNavLinks.length > 0 && (
            <div className="layout__item u-fit-sm">
                <nav className="navigation__container">
                    <ul className="navigation__nav-list">
                        {visibleNavLinks.map((navLink) => (
                            <li key={navLink.path}>
                                {(!navLink.subRoutes || navLink.subRoutes.length === 0) && (
                                    <NavLink
                                        to={navLink.path}
                                        activeClassName="active"
                                        className="navigation__nav-link"
                                    >
                                        <span className="label">{navLink.label}</span>
                                    </NavLink>
                                )}
                                {navLink.subRoutes && navLink.subRoutes.length > 0 && (
                                    <SubNavigationMenu
                                        menuLabel={navLink.label}
                                        active={location.pathname.startsWith(navLink.path)}
                                    >
                                        {navLink.subRoutes.map((subNavItem) => (
                                            <SubNavigationMenuItem
                                                key={subNavItem.path}
                                                path={subNavItem.path}
                                                label={subNavItem.label}
                                            />
                                        ))}
                                    </SubNavigationMenu>
                                )}
                            </li>
                        ))}
                    </ul>
                </nav>
            </div>
        )}

        {navLabel && (
            <div className="layout__item layout__item--right u-fit">
                <div className="navigation__label">
                    {navLabel}
                </div>
            </div>
        )}
    </div>
);

NavigationDesktop.displayName = 'NavigationDesktop';

export default withRouter(NavigationDesktop);
