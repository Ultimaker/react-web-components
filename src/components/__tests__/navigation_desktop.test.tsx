// Copyright (c) 2018 Ultimaker B.V.
import * as React from 'react';
import { shallow } from 'enzyme';

// component
import { NavigationDesktop } from '../navigation_desktop';

describe('The NavigationDesktop component', () => {
    let props;
    let wrapper;

    beforeEach(() => {
        props = {
            visibleNavLinks: [
                {
                    path: '/home',
                    label: 'Home',
                    visible: true,
                },
            ],
            location: {
                pathname: 'home',
            },
        };
        wrapper = shallow(<NavigationDesktop {...props} />);
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render navigation label', () => {
        wrapper.setProps({
            navLabel: 'Nav Label',
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should render sub navigation', () => {
        wrapper.setProps({
            visibleNavLinks: [
                {
                    path: '/home',
                    label: 'Home',
                    visible: true,
                    subRoutes: [
                        {
                            path: '/home/teams',
                            label: 'Teams',
                        },
                        {
                            path: '/home/printers',
                            label: 'Printers',
                        },
                    ],
                },
            ],
        });
        expect(wrapper).toMatchSnapshot();
    });
});
