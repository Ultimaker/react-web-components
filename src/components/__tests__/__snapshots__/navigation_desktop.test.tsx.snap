// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The NavigationDesktop component should render 1`] = `
<div
  className="navigation__desktop show-sm layout layout--gutter-none layout--align-middle"
>
  <div
    className="layout__item u-fit-sm"
  >
    <nav
      className="navigation__container"
    >
      <ul
        className="navigation__nav-list"
      >
        <li
          key="/home"
        >
          <NavLink
            activeClassName="active"
            aria-current="page"
            className="navigation__nav-link"
            to="/home"
          >
            <span
              className="label"
            >
              Home
            </span>
          </NavLink>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`The NavigationDesktop component should render navigation label 1`] = `
<div
  className="navigation__desktop show-sm layout layout--gutter-none layout--align-middle"
>
  <div
    className="layout__item u-fit-sm"
  >
    <nav
      className="navigation__container"
    >
      <ul
        className="navigation__nav-list"
      >
        <li
          key="/home"
        >
          <NavLink
            activeClassName="active"
            aria-current="page"
            className="navigation__nav-link"
            to="/home"
          >
            <span
              className="label"
            >
              Home
            </span>
          </NavLink>
        </li>
      </ul>
    </nav>
  </div>
  <div
    className="layout__item layout__item--right u-fit"
  >
    <div
      className="navigation__label"
    >
      Nav Label
    </div>
  </div>
</div>
`;

exports[`The NavigationDesktop component should render sub navigation 1`] = `
<div
  className="navigation__desktop show-sm layout layout--gutter-none layout--align-middle"
>
  <div
    className="layout__item u-fit-sm"
  >
    <nav
      className="navigation__container"
    >
      <ul
        className="navigation__nav-list"
      >
        <li
          key="/home"
        >
          <SubNavigationMenu
            active={false}
            menuLabel="Home"
          >
            <SubNavigationMenuItem
              key="/home/teams"
              label="Teams"
              path="/home/teams"
            />
            <SubNavigationMenuItem
              key="/home/printers"
              label="Printers"
              path="/home/printers"
            />
          </SubNavigationMenu>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
